libtransact/src/scheduler/multi/core.rs

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2019 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+//! Implementation of core MultiScheduler thread.
+
+use crate::scheduler::BatchExecutionResult;
+
+use std::sync::mpsc::Receiver;
+use std::sync::{Arc, Mutex};
+
+use super::shared::MultiSchedulerShared;
+
+/// An enum of messages which can be sent to the MultiSchedulerCore via a
+/// `Sender<MultiSchedulerCoreMessage>`
+pub enum MultiSchedulerCoreMessage {
+    /// The sub-scheduler at the given index in the sub-scheudlers list has gotten the given batch
+    /// result from the executor
+    BatchResult(usize, Option<BatchExecutionResult>),
+    /// An indicator to the `SchedulerCore` thread that it should exit its loop.
+    Shutdown,
+}
+
+pub struct MultiSchedulerCore {
+    shared_lock: Arc<Mutex<MultiSchedulerShared>>,
+    rx: Receiver<MultiSchedulerCoreMessage>,
+}
+
+impl MultiSchedulerCore {
+    pub fn new(
+        shared_lock: Arc<Mutex<MultiSchedulerShared>>,
+        rx: Receiver<MultiSchedulerCoreMessage>,
+    ) -> Self {
+        MultiSchedulerCore { shared_lock, rx }
+    }
+
+    fn run(&mut self) -> Result<(), String> {
+        loop {
+            // Receive batch result callbacks from the sub-schedulers
+            match self.rx.recv() {
+                Ok(MultiSchedulerCoreMessage::BatchResult(scheduler_index, batch_result)) => {
+                    let batch_result = batch_result.expect("Received batch result that is None");
+
+                    let mut shared = self
+                        .shared_lock
+                        .lock()
+                        .expect("scheduler shared lock is poisoned");
+                    let num_schedudlers = shared.schedulers().len();
+                    let pending_results = shared.pending_results_mut();
+
+                    // Get the pending result for this batch
+                    let (previous_result, already_executed_schedulers) = pending_results
+                        .get_mut(&batch_result.batch)
+                        .expect("Got callback for batch that's not pending");
```

Comment:
This could be handled via that error callback mechanism as well.

Reply:
Updated